deployment_service/app/deployment.py

```diff
@@ -0,0 +1,313 @@
+                self.logs,
+            )
+        return True
+
+    async def _health_check(self) -> bool:
+        """Performs a health check by making HTTP requests to a specified URL."""
+        self.logger.info(f"Performing health check on {self.healthcheck_url}...")
+        end_time = time.time() + HEALTH_CHECK_TIMEOUT
+        async with httpx.AsyncClient(timeout=2.0) as client:
+            while time.time() < end_time:
+                try:
+                    response = await client.get(self.healthcheck_url)
+                    self.logger.info(f"Health check response status: {response.status_code}")
+                    if response.status_code == 200:
+                        self.logger.info("Health check passed.")
+                        return True
+                except httpx.RequestError as exc:
+                    self.logger.warning(f"Health check request failed: {exc}")
+                await asyncio.sleep(HEALTH_CHECK_INTERVAL)
+        self.logger.error("Health check timed out.")
+        return False
+
+    def _cleanup(self, old_image_id: str) -> None:
+        """Removes the old Docker image after a successful deployment."""
+        if not old_image_id or old_image_id == self.image_tag:
+            return
+        try:
+            self.client.images.remove(old_image_id, force=True)
+            self.logger.info(f"Successfully removed old image: {old_image_id}")
+        except APIError:
+            self.logger.error("Could not remove old image. It may be in use.")
+
+    async def run(self) -> deque:
+        """Executes the full deployment flow with a temporary candidate of docker-compose.yml and atomic state update."""
+        self.logger.info(f"Starting deployment for commit {self.commit_sha}")
+        try:
+            candidate_dir = tempfile.mkdtemp(prefix=f"deploy-{self.commit_sha}")
+        except OSError as e:
+            raise CriticalError(f"Failed to create temporary directory: {e}", self.logs) from e
+
+        candidate_compose_path = os.path.join(
+            candidate_dir, os.path.basename(self.stable_compose_path)
+        )
+
+        try:
+            self.logger.info("Fetching latest data from git origin...")
+
+            if not await self._run_command(
+                ["git", "fetch", "origin"], cwd=self.stable_repository_dir
+            ):
+                raise CriticalError(
+                    "Failed to fetch from git origin. The commit might not be available.", self.logs
+                )
+            self.logger.info(
+                f"Copying candidate docker-compose.yml file at {candidate_dir} for commit {self.commit_sha}"
+            )
+            if not await self._run_command(
+                [
+                    "git",
+                    "show",
+                    f"{self.commit_sha}:{os.path.basename(self.stable_compose_path)}",
+                    ">",
+                    candidate_compose_path,
```

Comment:
The command executes bash? We can just pipe the output to the given file path, not need to use `>`.